### ISSUE TITLE
Crash during AI turn in Meduna

### ISSUE THREAD

**Issue by kavol** [ja2-stracciatella/ja2-stracciatella#495]
step one square with Ira, shoot the wounded enemy, hide her, end turn ... crash
```
(gdb) run
Starting program: /home/kavol/JA/ja2-stracciatella/ja2 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
warning: Missing auto-load script at offset 0 in section .debug_gdb_scripts
of file /home/kavol/JA/ja2-stracciatella/rust/debug/libstracciatella.so.
Use `info auto-load python-scripts [REGEXP]' to list them.
[New Thread 0x7ffff161a700 (LWP 2916)]
[New Thread 0x7ffff04a4700 (LWP 2917)]
2017/01/11 20:22:30 ERROR   [Music Control   ] Music Play Error -1 2
2017/01/11 20:22:40 WARNING [Tactical Save   ] SynchronizeItemTempFile() Reported 0, should be 31
2017/01/11 20:22:41 ERROR   [Music Control   ] Music Play Error -1 5
2017/01/11 20:22:45 ERROR   [Music Control   ] Music Play Error -1 4
2017/01/11 20:22:47 ERROR   [Music Control   ] Music Play Error -1 5
2017/01/11 20:22:51 ERROR   [Music Control   ] Music Play Error -1 2
2017/01/11 20:22:51 WARNING [Tactical Save   ] SynchronizeItemTempFile() Reported 0, should be 31
2017/01/11 20:22:52 ERROR   [Music Control   ] Music Play Error -1 5

Program received signal SIGSEGV, Segmentation fault.
0x0000000000762505 in CalcChanceToHitGun (pSoldier=0xe01ed0 <Menptr+86224>, sGridNo=11317, ubAimTime=4 '\004', ubAimPos=2 '\002')
    at /home/kavol/JA/ja2-stracciatella/src/game/Tactical/Weapons.cc:2158
2158                    if (GCM->getWeapon(usInHand)->ubWeaponClass == SMGCLASS)
(gdb) bt full
#0  0x0000000000762505 in CalcChanceToHitGun (pSoldier=0xe01ed0 <Menptr+86224>, sGridNo=11317, ubAimTime=4 '\004', ubAimPos=2 '\002') at /home/kavol/JA/ja2-stracciatella/src/game/Tactical/Weapons.cc:2158
        iChance = 79
        iRange = 150
        iSightRange = 0
        iMarksmanship = 79
        iPenalty = -20848
        ubAdjAimPos = 80 'P'
        tgt = 0x1900003300e4d800
        pTarget = 0x16006cffe01ed0
        iMaxRange = -20896
        bAttachPos = -1 '\377'
        iBonus = 1090702448
        usInHand = 151
        iScopeBonus = 51
        iGunCondition = 94
        pInHand = 0xe01f8c <Menptr+86412>
        bBandaged = -1 '\377'
        sDistVis = 0
#1  0x00000000007636cd in AICalcChanceToHitGun (pSoldier=0xe01ed0 <Menptr+86224>, sGridNo=11317, ubAimTime=4 '\004', ubAimPos=2 '\002') at /home/kavol/JA/ja2-stracciatella/src/game/Tactical/Weapons.cc:2630
        usTrueState = 1
        uiChance = 3349787243
#2  0x0000000000774e8b in CalcBestThrow (pSoldier=0xe01ed0 <Menptr+86224>, pBestThrow=0x7fffffffbac0) at /home/kavol/JA/ja2-stracciatella/src/game/TacticalAI/Attacks.cc:920
        ubLoop = 0 '\000'
        iOppThreatValue = {115, 0, 14676, 0, 1509, 0, 0, 0, 1, 32767, -159072064, 32767, 0, 0, -134225624, 32767, 50, 0, -134225624, 32767, 50, 0, -134230016, 32767, 8583658, 0, 0, 0, -19808, 32767, -161668538, 32767, -158974192, 32767, 1, 32767, -158974192, 32767, -19408, 32767, -19728, 32767, 50, 53, -19408, 
          32767, -19408, 32767, 9287220, 0, -18856, 32767, 16, 0, 0, 0, -18592, 32767, -162383979, 32767, -18592, 32767, -18592, 32767, -18592, 32767, -225829800, 32767, -18584, 32767, 1090748184, 0, 1090747836, 0, 0, 0, 0, 0, 1485569280, -945180053, -159064944, 32767, 1090747836, 0, 0, 0, -159065024, 32767, 8, 0, 
          1, 0, 32832, 0, 0, 0, -159065024, 32767, -162331269, 32767, 65536, 0, 118, 32767, 32816, 0, 32864, 0, 513, 0, 2052, 561, 64, 0, 8, 0, 155, 118, 1, 0, 0, 0, 120, 124, -19408, 32767, 165, 0, 10665270, 0, -19112, 32767, 165, 0, 1092472960, 0, -18864, 32767, -162390021, 32767, -72515583, 0, -18864, 32767, 
          -18864, 32767, -18864, 32767}
        sFriendTile = {7457, 7294, 5523, 9861, 9705, 0, 0, 0, 8, 0, 32767, 0, -14354, -2478, 32767, 0, -19088, -1, 32767, 0, 16, 0, 48, 0, -18896, -1, 32767, 0, -19088, -1, 32767, 0, 0, 0, 0, 0, -14354, -2478, 32767, 0, -19040, -1, 32767, 0, -19056, -1, 32767, 0, -17098, 162, 0, 0, -18744, -1, 32767, 0, 0, 0, 0, 0, 
          4499, -2480, 32767, 0, 0, 0, 32767, 0, -19328, -1, 0, 0, -20256, -1, 32767, 0, 0, 0, 32767, 0, -20216, -1, 32767, 0, -20232, -1, 32767, 0, 0, 0, 0, 0, 0, 0, 0, 0, -20208, -1, 32767, 0, 0, 0, 32767, 0, 0, 0, 32767, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32767, 0, -20192, -1, 32767, 0, 0, 0, 2, 0, 0, 0, 32767, 0, 
          0, 0, 0, 0, 8, 0, -1, -1, 18, 0, 32767, 0, -17094, 162, 0, 0}
        ubChanceToHit = 0 '\000'
        bPayloadPocket = 5 '\005'
        bYOffset = 0 '\000'
        bPersOL = 0 '\000'
        ubNumExcludedTiles = 1 '\001'
        bXOffset = 0 '\000'
        iTossRange = 50
        ubSafetyMargin = 6 '\006'
        iHitRate = 0
        iThreatValue = 115
        bFriendLevel = "\000\000\000\000\000\000\000\000\320\036\340\000\000\000\000\000P\271\377\377\377\177\000\000\320\036\340\000\000\000\000\000\063\000\000\000\000\000\000\000\300\314\002A\000\000\000\000@\271\377\377\377\177\000\000\337yR\000\000\000\000\000`\271\377\377\063\000\000\000 \225$A\000\000\000\000`\271\377\377\063\001\000\000\214\037\340\000\000\000\000\000p\271\377\377\063\000\000\000p\314\002A\000\000\000\004\240\271\377\377\377\177\000\000\265\234p\000\000\000\000\000\214\037\340\000\000\000\000\000Jes\000\v\000\000\000\240\271\377\377"
        fFriendsNearby = 0 '\000'
        usInHand = 51
        bEndLevel = 0 '\000'
        usGrenade = 0
        bMaxUp = 1 '\001'
        iEstDamage = 44
        ubMinAPcost = 10 '\n'
        uiPenalty = 32767
        ubChanceToGetThrough = 76 'L'
        ubChanceToReallyHit = 0 '\000'
        fSkipLocation = 0 '\000'
        bMaxRight = 1 '\001'
        bPublOL = 0 '\000'
        ubOpponentCnt = 1 '\001'
        ubRawAPCost = 10 '\n'
        usOppDist = 0
        ubOppsInRange = 1 '\001'
        iTotalThreatValue = 5060
        sGridNo = 11317
        bOpponentLevel = "\000M\222\000\063\000\000\000\020\000\000\000\000\000\000\000\020\272\377\377\377\177\000\000W\241p\000\000\000\000\000\000\271\377\377\001d\000\000\320\036\340\000\000\000\000\000\063\000\000\000\000\000\001\n\vJ\002A\000\000\063\000\214\037\340\000\000\000\000\000\337yR\000\000\000\000\000\063\000\000\000\000\000\000\000\300\314\002A\000\000\000\000\020\272\377\377\377\177\000\000\337yR\000\000\000\000\000 \272\377\377\063\000\000\000\000\375\213Xk\262\251\307@\272\377\377\377\177\000\000W\233p\000\000\000\000\000\000\272\377\377"
        opponents = {0xdee1d0 <Menptr+5072>, 0x7fffffffb662, 0xffffffffffffffff, 0x7fffffffb650, 0xffffffffffffffff, 0x0, 0x0, 0x0, 0x0, 0x0, 0x3e8, 0x100000000, 0x43b00000, 0x0, 0x8c0, 0x7ffff656df26 <__getdents64+86>, 0xffffffffffffffff, 0x0, 0xfffffffffffffea8, 0x2a, 0x7fffffffffff, 0x5185e0 <_start>, 
          0x7fffffffdf30, 0x7ffff684c900 <_IO_str_jumps>, 0x0, 0x7ffff656d9bd <closedir+13>, 0x1, 0xa5, 0x7fffffffb760, 0x10, 0xa6, 0x7ffff6508e97 <sprintf+135>, 0x0, 0x3000000020, 0x7fffffffb630, 0x7fffffffb570, 0x7fffffffb6f0, 0x7fffffffb6f0, 0x411ded18, 0x411de800, 0x0, 0x15, 0x43b01fc0, 0x0, 0xc7780000c780, 
          0x7fffffffba60, 0x7fffffffba60, 0x7fffffffba60, 0x7fffffffba88, 0x7fffffffbb9c, 0x7fffffffba60, 0x7fffffffbb9c, 0x0, 0x4, 0x895f70, 0x34, 0x1, 0x3100000006, 0x0, 0x82256f <CompareFileNames(void const*, void const*)+121>, 0x411dd080, 0x43ad52d8, 0x0, 0x411dd080, 0x6361667265746e49, 0x532e544345522f65, 
          0x49004954004954, 0x4f5252415f4d4f54, 0x5843502e5357, 0x7ffff652a12a <_IO_file_seekoff+522>, 0x8d80b4, 0x41015468, 0x7fffffffdf30, 0x900000000, 0x7ffff684dc90 <main_arena+80>, 0x7ffff6532748 <malloc+120>, 0x1c, 0x7ffff684dc40 <main_arena>, 0x10, 0x7ffff65945a9 <read+57>, 0x0, 
          0x7ffff652b878 <_IO_file_underflow+344>, 0x7ffff684dc40 <main_arena>, 0x411db3f0, 0x43d0c24a, 0x7ffff652aa53 <__GI__IO_file_xsgetn+275>, 0x1, 0x66c8, 0x411db3f0, 0x66c8, 0x0, 0x7ffff651ffd0 <fread+160>, 0x0, 0x66c8, 0x7fffffffb7a0, 0x5185e0 <_start>, 0x7fffffffdf30, 
          0x8222c2 <LoadDataFromLibrary(LibraryFile*, void*, unsigned int)+172>, 0x7ffff684dc90 <main_arena+80>, 0x66c8005185e0, 0x43d05f90, 0x7ffff684dc40 <main_arena>, 0x6, 0x441e3950, 0x66d0, 0x0, 0x7ffff684dc40 <main_arena>, 0x7ffff653057b <_int_malloc+379>, 0x48, 0x74, 0x66c8, 0x66f0, 0x19b, 0x1cb0000066d, 
          0x33, 0x21, 0x200, 0x230, 0x8, 0x3800000021, 0x1, 0x0, 0x6e0000005c, 0x0, 0x0, 0x7c00000077, 0x6e0000005d, 0x7ffff684dc40 <main_arena>, 0x200, 0x5185e0 <_start>, 0x7fffffffdf30, 0x430410b0, 0x7fffffffb8c0, 0x836344 <SGPVObject::SubregionProperties(unsigned long) const+46>, 0x27, 0x430410b0, 0x7fffffffb8d0, 
          0x8268ee <XMalloc(unsigned long)+25>, 0x7fffffffb8c0, 0xc7a9b26b588bfd00, 0x43049560, 0x430410b0, 0x7fffffffb960, 0x837a6f <Blt8BPPDataTo16BPPBufferMonoShadowClip(unsigned short*, unsigned int, SGPVObject*, int, int, unsigned short, SGPRect*, unsigned short, unsigned short, unsigned short)+616>, 
          0x1ffffb920, 0xffff00830000, 0x300000027, 0x430410b0}
        ubSearchRange = 1 '\001'
        sExcludeTile = {11317, 0 <repeats 99 times>}
        ubLoop2 = 1 '\001'
        iAttackValue = 0
        ubMaxPossibleAimTime = 4 '\004'
        bMaxDown = 1 '\001'
        ubDiff = 3 '\003'
        sEndGridNo = 0
        sOpponentTile = {11317, -1, 32767, 0, 0, 0, 0, 0, -1, -1, -1, -1, 0, 0, 2, 0, 0, 0, 0, 0, -17094, 162, 0, 0, 0, 0, 32767, 0, 16, 0, 48, 0, -18576, -1, 32767, 0, -18768, -1, 32767, 0, 0, 0, 114, 0, 0, 0, 0, 0, 11, 0, 101, 0, 16, 0, 48, 0, -18528, -1, 32767, 0, -18720, -1, 32767, 0, -19984, -1, 32767, 0, 0, 0, 
          0, 0, -1, -1, -1, -1, 0, 0, 2, 0, 32, 0, 115, 0, -17094, 162, 0, 0, 110, 0, 100, 0, 32, 0, 111, 0, 102, 0, 32, 0, 77, 0, 79, 0, -21053, 165, 0, 0, -20032, -1, 32767, 0, -3528, 171, 0, 0, 15650, -2426, 32767, 0, -32160, 81, 0, 0, -8, -1, -1, -1, -7864, 163, 0, 0, -19680, -1, 32767, 0, -3528, 171, 0, 0, 
          17260, -2426, 32767, 0, 27224, -2426, 32767, 256}
        ubFriendCnt = 5 '\005'
        ubOppsAdjacent = 1 '\001'
        bMaxLeft = 1 '\001'
#3  0x0000000000776bc6 in CheckIfTossPossible (pSoldier=0xe01ed0 <Menptr+86224>, pBestThrow=0x7fffffffbac0) at /home/kavol/JA/ja2-stracciatella/src/game/TacticalAI/Attacks.cc:1695
        ubMinAPcost = 10 '\n'
#4  0x000000000077c775 in DecideActionRed (pSoldier=0xe01ed0 <Menptr+86224>, ubUnconsciousOK=1 '\001') at /home/kavol/JA/ja2-stracciatella/src/game/TacticalAI/DecideAction.cc:1713
        iDummy = 32767
        bHighestWatchLoc = 0 '\000'
        sClosestOpponent = 0
        sDistVisible = -1
        bInWater = 0 '\000'
        BestThrow = {opponent = 0x0, ubPossible = 0 '\000', ubAimTime = 0 '\000', ubChanceToReallyHit = 0 '\000', iAttackValue = 0, sTarget = 25601, bTargetLevel = -1 '\377', ubAPCost = 0 '\000', bWeaponIn = 5 '\005'}
        bSeekPts = 0 '\000'
        bHelpPts = 0 '\000'
        bHidePts = 0 '\000'
        bActionReturned = 0 '\000'
        bInDeepWater = 0 '\000'
        bInGas = 0 '\000'
        bWatchPts = 0 '\000'
        fClimb = 0 '\000'
        fCivilian = 0 '\000'
        ubCanMove = 1 '\001'
        iChance = -17696
        sClosestFriend = 32767
        sClosestDisturbance = -17624
        sCheckGridNo = -1
        in_light_at_night = false
#5  0x00000000007809c4 in DecideAction (pSoldier=0xe01ed0 <Menptr+86224>) at /home/kavol/JA/ja2-stracciatella/src/game/TacticalAI/DecideAction.cc:3667
        bAction = 0 '\000'
#6  0x000000000076d565 in TurnBasedHandleNPCAI (pSoldier=0xe01ed0 <Menptr+86224>) at /home/kavol/JA/ja2-stracciatella/src/game/TacticalAI/AIMain.cc:1175
No locals.
#7  0x000000000076c45d in HandleSoldierAI (pSoldier=0xe01ed0 <Menptr+86224>) at /home/kavol/JA/ja2-stracciatella/src/game/TacticalAI/AIMain.cc:407
No locals.
#8  0x00000000006f9d68 in ExecuteOverhead () at /home/kavol/JA/ja2-stracciatella/src/game/Tactical/Overhead.cc:1111
        pSoldier = 0xe01ed0 <Menptr+86224>
        soldier = {px = 0x43a8ae40, pn = {pi_ = 0x43a8ae20}}
        cnt = 15
        iTimerVal = 1341300180
        fHandleAI = 0 '\000'
        iTimerTest = 1341300180
        fKeepMoving = 88 'X'
#9  0x00000000007f5164 in MainGameScreenHandle () at /home/kavol/JA/ja2-stracciatella/src/game/GameScreen.cc:494
No locals.
#10 0x00000000007f36d0 in GameLoop () at /home/kavol/JA/ja2-stracciatella/src/game/GameLoop.cc:268
        InputEvent = {usKeyState = 0, usEvent = 32, usParam = 0, Char = 0 L'\000'}
        uiOldScreen = GAME_SCREEN
        MousePos = {iX = 287, iY = 249}
#11 0x0000000000828f9f in MainLoop (msPerGameCycle=25) at /home/kavol/JA/ja2-stracciatella/src/sgp/SGP.cc:239
        gameCycleMS = 45427
        event = {type = 769, common = {type = 769, timestamp = 42887}, window = {type = 769, timestamp = 42887, windowID = 2, event = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000', padding3 = 0 '\000', data1 = 7, data2 = 100}, key = {type = 769, timestamp = 42887, windowID = 2, state = 0 '\000', 
            repeat = 0 '\000', padding2 = 0 '\000', padding3 = 0 '\000', keysym = {scancode = SDL_SCANCODE_D, sym = 100, mod = 0, unused = 3349787243}}, edit = {type = 769, timestamp = 42887, windowID = 2, 
            text = "\000\000\000\000\a\000\000\000d\000\000\000\000\000\213Xk\262\251\307\000\000\000\000\000\000\000\000\300\342\377@", start = 0, length = 0}, text = {type = 769, timestamp = 42887, windowID = 2, 
            text = "\000\000\000\000\a\000\000\000d\000\000\000\000\000\213Xk\262\251\307\000\000\000\000\000\000\000\000\300\342\377@"}, motion = {type = 769, timestamp = 42887, windowID = 2, which = 0, state = 7, x = 100, y = 1485504512, xrel = -945180053, yrel = 0}, button = {type = 769, timestamp = 42887, 
            windowID = 2, which = 0, button = 7 '\a', state = 0 '\000', clicks = 0 '\000', padding1 = 0 '\000', x = 100, y = 1485504512}, wheel = {type = 769, timestamp = 42887, windowID = 2, which = 0, x = 7, y = 100, direction = 1485504512}, jaxis = {type = 769, timestamp = 42887, which = 2, axis = 0 '\000', 
            padding1 = 0 '\000', padding2 = 0 '\000', padding3 = 0 '\000', value = 7, padding4 = 0}, jball = {type = 769, timestamp = 42887, which = 2, ball = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000', padding3 = 0 '\000', xrel = 7, yrel = 0}, jhat = {type = 769, timestamp = 42887, which = 2, 
            hat = 0 '\000', value = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000'}, jbutton = {type = 769, timestamp = 42887, which = 2, button = 0 '\000', state = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000'}, jdevice = {type = 769, timestamp = 42887, which = 2}, caxis = {type = 769, timestamp = 42887, 
            which = 2, axis = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000', padding3 = 0 '\000', value = 7, padding4 = 0}, cbutton = {type = 769, timestamp = 42887, which = 2, button = 0 '\000', state = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000'}, cdevice = {type = 769, timestamp = 42887, which = 2}, 
          adevice = {type = 769, timestamp = 42887, which = 2, iscapture = 0 '\000', padding1 = 0 '\000', padding2 = 0 '\000', padding3 = 0 '\000'}, quit = {type = 769, timestamp = 42887}, user = {type = 769, timestamp = 42887, windowID = 2, code = 0, data1 = 0x6400000007, data2 = 0xc7a9b26b588b0000}, syswm = {
            type = 769, timestamp = 42887, msg = 0x2}, tfinger = {type = 769, timestamp = 42887, touchId = 2, fingerId = 429496729607, x = 1.22265693e+15, y = -86884.8359, dx = 0, dy = 0, pressure = 7.99642944}, mgesture = {type = 769, timestamp = 42887, touchId = 2, dTheta = 9.80908925e-45, dDist = 1.40129846e-43, 
            x = 1.22265693e+15, y = -86884.8359, numFingers = 0, padding = 0}, dgesture = {type = 769, timestamp = 42887, touchId = 2, gestureId = 429496729607, numFingers = 1485504512, error = -86884.8359, x = 0, y = 0}, drop = {type = 769, timestamp = 42887, 
            file = 0x2 <error: Cannot access memory at address 0x2>}, padding = "\001\003\000\000\207\247\000\000\002\000\000\000\000\000\000\000\a\000\000\000d\000\000\000\000\000\213Xk\262\251\307\000\000\000\000\000\000\000\000\300\342\377@", '\000' <repeats 11 times>}
        s_doGameCycles = 1 '\001'
#12 0x0000000000829a4b in main (argc=1, argv=0x7fffffffdf38) at /home/kavol/JA/ja2-stracciatella/src/sgp/SGP.cc:474
        params = 0x40ffd9d0
        configFolderPath = "/home/kavol/.ja2"
        exeFolder = "/home/kavol/JA/ja2-stracciatella"
        rustConfigFolderPath = 0x41150550 "FONT12ARIAL.STI"
        gameResRootPath = "/home/kavol/JA/JA2_GOLD/"
        extraDataDir = "/home/kavol/JA/ja2-stracciatella"
        cm = 0x4102cc70
        libraries = std::vector of length 21, capacity 32 = {"Fonts.slf", "Radarmaps.slf", "Music.slf", "Anims.slf", "BattleSNDS.slf", "Faces.slf", "Bigitems.slf", "Speech.slf", "Tilesets.slf", "Mercedt.slf", "Interface.slf", "Cursors.slf", "Npc_speech.slf", "Npcdata.slf", "Maps.slf", "Loadscreens.slf", 
          "Laptop.slf", "Data.slf", "Ambient.slf", "Sounds.slf", "Binarydata.slf"}
        result = true
        version = GV_ENGLISH
        rustResRootPath = 0x41017fa0 "A1.STI"
        externalizedDataPath = "/home/kavol/JA/ja2-stracciatella/externalized"
```

[QS-Meduna.zip](https://github.com/ja2-stracciatella/ja2-stracciatella/files/768799/QS-Meduna.zip)

**Comment by JAsmine-ja2** [comment]
I can confirm this crash on JA2 Stracciatella 0.15.1
Move Ira one step in NE direction, shoot the soldier, hide her... crash.

On Vanilla JA2 1.12 (same JA2 as above, just without Stracc), things look different: The game doesn't crash, but isn't able to create an AUTOsave with autosave.pls file:
![screen000](https://user-images.githubusercontent.com/17213795/28705034-389e6062-736e-11e7-8901-223631f18a82.jpg)

**Comment by Manuel-K** [comment]
I've assigned this issue to me as a reminder to myself…

**Comment by Manuel-K** [comment]
It still crashes on current master.

**Comment by Manuel-K** [comment]
As the above stack trace tells us, the problem is here: https://github.com/ja2-stracciatella/ja2-stracciatella/blob/master/src/game/Tactical/Weapons.cc#L2169


`GCM->getWeapon(pSoldier->usAttackingWeapon)` is `NULL`.

Preventing the segmentation fault is trivial.

The attacker ("SOLDIER") has `usAttackingWeapon == 151` which is `SMOKE_GRENADE` and that is not a `WeaponModel`. It's just an `ItemModel` and obviously `isTwoHanded()` is false.


I do not understand the that code block at all. The logic seems to be:
If we do not use a two-handed weapon, we assume that we are dual-wielding and:
- give a penalty to SMGs.
- if the soldier is not ambidextrous and `!IsValidSecondHandShot()` is true penalize the hit chance by `AIM_PENALTY_DUAL_PISTOLS`.

The second part seems correct, because `IsValidSecondHandShot()` checks if the soldier has two weapons. 
The first part does not do that. The `TODO0007` note mentions this as well: “Deduct AIM_PENALTY_SMG when firing a SMG even when second hand is free?”

**Comment by Manuel-K** [comment]
Moved the rest to #700.

**Comment by chmallon** [comment]
I think the patch is wrong.

> The attacker ("SOLDIER") has usAttackingWeapon == 151 which is SMOKE_GRENADE

It makes no sense to look at a `SMOKE_GRENADE` in `CalcChanceToHitGun()`(!) in the first place.

The (indirect) caller looks like this:
```c++
        if ( EXPLOSIVE_GUN( usInHand ) )
        {
          ubRawAPCost   = MinAPsToShootOrStab(*pSoldier, sGridNo, FALSE);
          ubChanceToHit = (UINT8) AICalcChanceToHitGun(pSoldier, sGridNo, ubMaxPossibleAimTime, AIM_SHOT_TORSO );
        }
```

The definition of `EXPLOSIVE_GUN()` is:
```c++
#define EXPLOSIVE_GUN( x )   ( x == ROCKET_LAUNCHER || x == TANK_CANNON )
```

I.e. the call should be looking at a `ROCKET_LAUNCHER` or a `TANK_CANNON` and definitely not a `SMOKE_GRENADE`.
Probably `pSoldier->usAttackingWeapon` is not correctly set.
This is a typical result of accumulate and fire.

`usInHand` in the caller is always `usInHand = pSoldier->inv[HANDPOS].usItem`.
But then in `CalcChanceToHitGun()` the code looks at `pInHand = &(pSoldier->inv[pSoldier->ubAttackingHand])` and `usInHand = pSoldier->usAttackingWeapon`.
Probably both `ubAttackingHand` and `usAttackingWeapon` are not set correctly.

**Comment by Manuel-K** [comment]
>I think the patch is wrong.

Checking for null is never wrong.

If you want to fix the underlying issue in addition to the direct problem, feel free to create a PR.

**Comment by chmallon** [comment]
> Checking for null is never wrong.

This is an incorrect assumption.
Checking it there just silently hides an underlying problem.
There should be no null pointer there in the first place.
Therefore no kind of "gracefully" handling the null pointer there can be correct.
As a result the program just limps on with incorrect data which in turn may cause further hard to diagnose problems later on.

In the past JA2 had a lot more places which silently did something strange.
Due to internal restructuring -- usually by changing from indices to pointers -- a lot of them were turned from silently corrupting the game state, which is nigh impossible to trace back to the underlying cause, into a crash, which is a more distinct symptom and usually way closer to the cause of the problem.
This helped to improve the stability in the long run.

**Comment by Manuel-K** [comment]
@chmallon 

Okay. You seriously need to adapt a friendlier attitude or start to think what you write. You always write everything from the point that only you can be right and everyone else is wrong. I am not getting payed for spending so much time on this project.

Everything you write is formulated in a condescending way. You are definitely Linus Torvalds and he only barely gets away with it. This makes the content almost meaningless.

I'm seriously thinking about stopping any work on this project.

**Comment by chmallon** [comment]
> You are definitely Linus Torvalds

Wow, I am seriously offended by the comparison of this person with me.
He regularly insults people by ad hominem, which I clearly not do.
I simply state when I think something is wrong and also explain the technical reason for this.
I am not hiding this behind flowery words of so called political correctness, but there is a huge difference between stating something clearly as well as unambiguously and insulting somebody at a personal level.
Sadly one of Linus Torvalds' dubious achievements is blurring the line between these.

I also consider it very disrespectful that you did not write me this in a personal message but immediately made it public here.
This way you forced me to also respond here in public and further do so in a thread which has nothing to this.
There are several more points to make, but this is too much off-topic already.
I try to make solid technical contributions and as response get insulted -- I am very disappointed.

**Comment by lynxlynxlynx** [comment]
Maybe you're not aware of it, but you *do* sometimes come across as a bit crass, whether it's intentional or not. In this case, the patch is not wrong, it just isn't ideal, since the root cause remains unknown and will be harder to detect. Do we need to care about it? Maybe. Maybe not.

In any case, the improvement can easily be a separate issue (I'll go create it now) and in the future please just open them as such.

**Comment by derSteFfi** [comment]
> You are definitely Linus Torvalds
While you are kind of right, @chmallon tone could be friendlier, this isn't helpful, too. And not correct in this case, Christoph never told you to go f#ck youself or similar. We all do this for fun in our free time, so we should stick to a friendly attitude, yes. On the other hand, we all do this to learn stuff and getting mistakes or misassumptions pointed out can come across harsh and friendlyness is secondary in that matter.

In short: You both have a point. We're adults, let's stick to the interesting part.

**Comment by kavol** [comment]
guys, no need to overreact ...
as it was quoted before:
> I think the patch is wrong.

not sure what does it mean in the writer's and every possible reader's cultural context, but where I am from, saying "I think ..." conjures up doubt
it is not the same as "I know" or "I'm sure" or so - it implicitly denies reading the following as "everyone else is wrong", it rather asks for explanation what's wrong about author's POV and arguments
and it doesn't need to be repeated in subsequent comment as the comment just expands on explaining the POV that led to the first one

on the other hand, I find it a bit schizophrenic to reject political correctness and then complain about things being discussed in public - again, in my place, being "politicaly incorrect" means telling things straight **and** in public

while I tend to disagree with @chmallon, I don't like the tone on **both** sides, so despite not exactly constructive for resolving the code issues[*], I totally support voicing the concerns about the communication, now, before the frustration leads to losing valuable contributors ... **we all** have some room to improve, and it is not just about "flowery words" but also e.g. about being too quick in own judgements

[*] thanks @lynxlynxlynx meanwhile - I was too slow writing this :-)

**Comment by Manuel-K** [comment]
>  You are definitely Linus Torvalds …
There is a not missing. I'm sorry about that, but I'm fed up with this.

Well, it does not matter anymore. I don't want to waste any more time on this at all.

**Comment by Manuel-K** [comment]
> I totally support voicing the concerns about the communication, now, before the frustration leads to losing valuable contributors

Too late.

**Comment by JAsmine-ja2** [comment]
We all support this project for fun and in our free time. It's our hobby and, to some degree, our passion.
Online communities are what they are and often it's not easy to communitcate with just some lines of text, maybe even in a language that is not your first language. Misunderstandings (failures to communicate) happen all the time, especially online.

Being criticised for what you voluntarily do in your free time doesn't feel right at all.
On the other hand, since it's just a hobby, no one can (or does) expect perfection and if something you do is not perfect, or even wrong - it's not so bad and there are no consequences to fear.

The way I see it,
@chmallon pointed out some technical issues and was very blunt about it.
@Manuel-K on the other hand pointed out concerns about the communication and was very blunt about it.

Calm down, stop the hate, participate ❤

**Comment by Manuel-K** [comment]
I'm also disappointed that nobody noticed my typo – without it the whole sentence is complete garbage.

It obviously that my wording was harsh, but it was an answer to a two digit number of harsh messages by one person. A person that has proven before that it's pointless to write anything in a friendly way, because everything gets ignored anyway. Remember the time that he tried to stop me from fixing a bug that was caused by a design decision (that I was not involved in) that he did not like. I was attacked because I would not override that earlier decision without any input from the decision makers. He ignored any reasoning that I would not go behind their backs… 

The point that @derSteFfi does not want to see (for the second time) is that I would not accept that behaviour even if I was payed to do this work. Obviously I'm not getting payed, so I will cease the work.

I had an assert in the code in question, but I removed it because it is utterly pointless. The code base has so many other problems that it does not matter if there is a minor problem in the AI code that does not stop the AI from firing the rocket launcher at Ira.

I'm convinced that a game that is so easy to crash is not a game that is meant to be played. In the first fifteen minutes of my first contact with this project, I've run into the misnamed `data_dir` issue (i.e. do not point to data although it says so) – there is no helpful error message. Yeah, you can see something if you run the game in a debugger, but that is not useful at all. Then I've encountered the laptop crash during the first five minutes in the game. That's a horrible first impression. I assume that most people will stop at this point. The number of crashes that do not get reported has to be much larger than what we are seeing here. There is no graceful handling of anything that is wrong. That's not the way to do things if you have non developers as users. I would not use software that acts like that.  

I've clearly made the mistake that it is meant to be played. 


I'm not willing to spend any more time on this. The way this is going is no fun. 

On top of the possibility of segmentation faults, there are some things that I as a user would need and I was reviewing earlier this week before I would consider playing the game with this project is a working scaling system. As it stands now, the option for higher resolutions is not much use except on the tactical screen. There has to be some internal scalers that try to use the highest possible integer scaling factor that makes sense for the current screen. The strategic screen is horrible at any resolution higher than the original resolutions (”Oh, look, huge black borders…”). The bottom bar on the tactical screen should have a scaling option that is separate from the tile view. OpenXcom does this in an excellent way. Obviously scaling would increase the drawing overhead (the scaling itself and additional `SDL_BlitSurface()` calls), but it is import. This brings us to another problem: The license situation. It makes developing much harder than necessary. This complicates the inclusion of the UB code as well. That work was a huge waste of my time.

So, no, I'd rather do something else if I'm expected to accept this communication behaviour. Today was the second time a played a game this year.

This is probably full of typos that everyone will take as insults, but I can't be bothered to check the text.

@selaux @lynxlynxlynx @JAsmine-ja2: It was nice working with you. Thank you.

**Comment by lynxlynxlynx** [comment]
Likewise Manuel, it's sad it had come to this, but it is what it is.

Regarding scaling, window mode + window manager resize/fullscreen works fine; I did the last game that way. To me it doesn't even have to be an integer factor to look ok.
I also think that in the short history of this project we removed many more crashes than added, but I don't really want to derail further.

So, what can we learn from this? It's obvious we don't have a good place to discuss any of this, so I'll open a separate issue. And personally, to try to be more empathic.
